Altair: Using incorrect state to calculate sync committee duties

### Description
With minimal configuration and Altair fork slot set to 8, sometimes but not all the time I get the exception:

```
2021-05-10 14:50:27.016 | validator-async-2         | ERROR | RetryingDutyLoader             | Failed to request validator duties for epoch 7. Retrying after delay.
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: State must be in the same or previous sync committee period. Cannot calculate epoch {} from state at slot {} [7, 0]
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:683) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658) ~[?:?]
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2158) ~[?:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.thenApply(SafeFuture.java:408) ~[classes/:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.getSyncCommitteeDuties(ValidatorApiHandler.java:217) ~[classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:564) ~[?:?]
	at tech.pegasys.teku.infrastructure.events.DirectEventDeliverer.executeMethod(DirectEventDeliverer.java:74) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.events.DirectEventDeliverer.deliverToWithResponse(DirectEventDeliverer.java:67) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.events.AsyncEventDeliverer.lambda$deliverToWithResponse$1(AsyncEventDeliverer.java:75) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.events.AsyncEventDeliverer$QueueReader.deliverNextEvent(AsyncEventDeliverer.java:117) ~[classes/:?]
	at tech.pegasys.teku.infrastructure.events.AsyncEventDeliverer$QueueReader.run(AsyncEventDeliverer.java:109) ~[classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
	at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: java.lang.IllegalArgumentException: State must be in the same or previous sync committee period. Cannot calculate epoch {} from state at slot {} [7, 0]
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:441) ~[guava-29.0-jre.jar:?]
	at tech.pegasys.teku.spec.logic.common.util.SyncCommitteeUtil.getSyncSubcommittees(SyncCommitteeUtil.java:100) ~[classes/:?]
	at tech.pegasys.teku.spec.logic.common.util.SyncCommitteeUtil.getCommitteeIndices(SyncCommitteeUtil.java:194) ~[classes/:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$getSyncCommitteeDuty$33(ValidatorApiHandler.java:672) ~[classes/:?]
	at java.util.Optional.map(Optional.java:258) ~[?:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.getSyncCommitteeDuty(ValidatorApiHandler.java:672) ~[classes/:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$getSyncCommitteeDutiesFromIndexesAndState$32(ValidatorApiHandler.java:662) ~[classes/:?]
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:271) ~[?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.getSyncCommitteeDutiesFromIndexesAndState(ValidatorApiHandler.java:663) ~[classes/:?]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$getSyncCommitteeDuties$7(ValidatorApiHandler.java:220) ~[classes/:?]
	at java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:680) ~[?:?]
	... 16 more
```

--- a/ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtil.java
+++ b/ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtil.java
@@ -284,7 +284,10 @@ public class SyncCommitteeUtil {
     // Can look-ahead one sync committee period.
     final UInt64 requiredSyncCommitteePeriod =
         computeSyncCommitteePeriod(requiredEpoch).minusMinZero(1);
-    return requiredSyncCommitteePeriod.times(specConfig.getEpochsPerSyncCommitteePeriod());
+    return requiredSyncCommitteePeriod
+        .times(specConfig.getEpochsPerSyncCommitteePeriod())
+        // But can't use a state from before the Altair fork
+        .max(miscHelpers.computeEpochAtSlot(specConfig.getAltairForkSlot()));
   }
 
   public UInt64 computeFirstEpochOfCurrentSyncCommitteePeriod(final UInt64 currentEpoch) {

--- a/ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtilTest.java
+++ b/ethereum/spec/src/test/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtilTest.java
@@ -156,6 +156,18 @@ class SyncCommitteeUtilTest {
     assertMinEpochForSyncCommitteePeriod(5);
   }
 
+  @Test
+  void getMinEpochForSyncCommitteeAssignments_shouldNotAllowEpochToBeLessThanForkEpoch() {
+    final UInt64 altairForkEpoch = UInt64.ONE;
+    final UInt64 altairForkSlot = spec.computeStartSlotAtEpoch(altairForkEpoch);
+    final Spec spec = TestSpecFactory.createMinimalWithAltairFork(altairForkSlot);
+    final SyncCommitteeUtil syncCommitteeUtil = spec.getSyncCommitteeUtilRequired(altairForkSlot);
+    assertThat(syncCommitteeUtil.getMinEpochForSyncCommitteeAssignments(altairForkEpoch))
+        .isEqualTo(altairForkEpoch);
+    assertThat(syncCommitteeUtil.getMinEpochForSyncCommitteeAssignments(UInt64.valueOf(2)))
+        .isEqualTo(altairForkEpoch);
+  }
+
   private void assertMinEpochForSyncCommitteePeriod(final int period) {
     final UInt64 epochsPerPeriod = UInt64.valueOf(config.getEpochsPerSyncCommitteePeriod());
     UInt64.range(epochsPerPeriod.times(period), epochsPerPeriod.times(period + 1))

--- a/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java
+++ b/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java
@@ -641,7 +641,7 @@ public class ValidatorApiHandler implements ValidatorApiChannel {
     final UInt64 stateEpoch = spec.getCurrentEpoch(bestState);
     if (epoch.isGreaterThan(stateEpoch)) {
       // Use the earliest possible epoch since we'll need to process empty slots
-      requiredEpoch = syncCommitteeUtil.computeFirstEpochOfCurrentSyncCommitteePeriod(epoch);
+      requiredEpoch = syncCommitteeUtil.getMinEpochForSyncCommitteeAssignments(epoch);
     } else {
       // Use the latest possible epoch since it's most likely to still be in memory
       requiredEpoch = syncCommitteeUtil.computeLastEpochOfCurrentSyncCommitteePeriod(epoch);

--- a/validator/coordinator/src/test/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandlerTest.java
+++ b/validator/coordinator/src/test/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandlerTest.java
@@ -89,6 +89,7 @@ import tech.pegasys.teku.validator.api.ProposerDuties;
 import tech.pegasys.teku.validator.api.ProposerDuty;
 import tech.pegasys.teku.validator.api.SendSignedBlockResult;
 import tech.pegasys.teku.validator.api.SubmitCommitteeSignatureError;
+import tech.pegasys.teku.validator.api.SyncCommitteeDuties;
 import tech.pegasys.teku.validator.api.SyncCommitteeSubnetSubscription;
 import tech.pegasys.teku.validator.coordinator.performance.DefaultPerformanceTracker;
 
@@ -363,6 +364,44 @@ class ValidatorApiHandlerTest {
                 firstSlotAfterNextSyncCommitteePeriod, List.of(1)))
         .isCompletedExceptionallyWith(IllegalArgumentException.class)
         .hasMessageContaining("not within the current or next sync committee periods");
+  }
+
+  @Test
+  void getSyncCommitteeDuties_shouldNotUseEpochPriorToFork() {
+    final Spec spec = TestSpecFactory.createMinimalWithAltairFork(EPOCH_START_SLOT);
+    final ValidatorApiHandler validatorApiHandler =
+        new ValidatorApiHandler(
+            chainDataProvider,
+            chainDataClient,
+            syncStateProvider,
+            blockFactory,
+            blockImportChannel,
+            blockGossipChannel,
+            attestationPool,
+            attestationManager,
+            attestationTopicSubscriptions,
+            activeValidatorTracker,
+            dutyMetrics,
+            performanceTracker,
+            spec,
+            forkChoiceTrigger,
+            syncCommitteeSignaturePool,
+            syncCommitteeContributionPool,
+            syncCommitteeSubscriptionManager);
+    // Best state is still in Phase0
+    final BeaconState state =
+        dataStructureUtil.stateBuilderPhase0().slot(PREVIOUS_EPOCH_START_SLOT.minus(1)).build();
+    when(chainDataClient.getCurrentEpoch()).thenReturn(EPOCH);
+    when(chainDataClient.getBestState()).thenReturn(Optional.of(state));
+    when(chainDataClient.getStateAtSlotExact(any())).thenReturn(new SafeFuture<>());
+
+    final SafeFuture<Optional<SyncCommitteeDuties>> result =
+        validatorApiHandler.getSyncCommitteeDuties(EPOCH, List.of(1));
+    assertThat(result).isNotDone();
+
+    // The start of the sync committee period is prior to the fork block so we should use the
+    // fork block to ensure we actually have sync committees available.
+    verify(chainDataClient).getStateAtSlotExact(EPOCH_START_SLOT);
   }
 
   @Test

